extensions/amp-viewer-integration/0.1/messaging.js

```diff
@@ -66,123 +86,174 @@ export class Messaging {
+  sendResponseError_(requestId, messageName, reason) {
+    this.logError_(
+      TAG + ': sendResponseError_, Message name: ' + messageName, reason);
+    this.sendMessage_({
+      app: APP,
+      requestid: requestId,
+      type: MessageType.RESPONSE,
+      name: messageName,
+      data: null,
+      rsvp: undefined,
+      error: reason,
+    });
   }
 
   /**
-   * @param {string} requestId
-   * @param {*} reason
+   * @param {Message} message
    * @private
    */
-  sendResponseError_(requestId, reason) {
-    this.sendMessage_(
-      responseSentinel_, requestId, 'ERROR', reason, false);
+  sendMessage_(message) {
+    this.target_./*OK*/postMessage(message, this.targetOrigin_);
   }
 
   /**
    * I'm handing an incoming request from Bob. I'll either respond normally
    * (ex: "got it Bob!") or with an error (ex: "I didn't get a word of what
    * you said!").
-   * @param {*} message
+   * @param {Message} message
    * @private
    */
   handleRequest_(message) {
-    dev().info(TAG, 'messaging.js -> handleRequest_');
-    const requestId = message.requestId;
-    const promise = this.requestProcessor_(message.type, message.payload,
-        message.rsvp);
+    dev().info(TAG, 'handleRequest_', message);
+    if (!this.requestProcessor_) {
+      throw new Error(
+        'Cannot handle request because handshake is not yet confirmed!');
+    }
+    const requestId = message.requestid;
+    const msg = dev().assertString(message.name);
     if (message.rsvp) {
+      const promise =
+        this.requestProcessor_(msg, message.data, message.rsvp);
       if (!promise) {
-        this.sendResponseError_(requestId, 'no response');
+        this.sendResponseError_(requestId, msg, 'no response');
         dev().assert(promise,
-          'expected response but none given: ' + message.type);
+          'expected response but none given: ' + message.name);
       }
-      promise.then(payload => {
-        this.sendResponse_(requestId, payload);
+      promise.then(data => {
+        this.sendResponse_(requestId, msg, data);
       }, reason => {
-        this.sendResponseError_(requestId, reason);
+        this.sendResponseError_(requestId, msg, /** @type {string} */(reason));
       });
     }
   }
 
   /**
    * I sent out a request to Bob. He responded. And now I'm handling that
    * response.
-   * @param {*} message
+   * @param {Message} message
    * @private
    */
   handleResponse_(message) {
-    dev().info(TAG, 'messaging.js -> handleResponse_');
-    const requestId = message.requestId;
+    dev().info(TAG, 'handleResponse_');
+    const requestId = message.requestid;
     const pending = this.waitingForResponse_[requestId];
     if (pending) {
       delete this.waitingForResponse_[requestId];
-      if (message.type == 'ERROR') {
-        pending.reject(message.payload);
+      if (message.error) {
+        this.logError_(TAG + ': handleResponse_ error: ', message.error);
+        pending.reject(message.error);
```

Comment:
This should be `pending.reject(new Error(message.error))`. This is a very important nuance and the one we are trying to address on compiler level as well. In the promise land - we only operate with `Error` objects. But in messaging, we have to encode the error somehow, e.g. as a string. These conversions should be very explicit in our code as well.